### ISSUE TITLE
Elections can hang when electorate grows 

### ISSUE THREAD

**Issue by adamhjk** [habitat-sh/habitat#1994]
# tl;dr

Elections in habitat require a quorum of alive members in order to reach completion. When members join a service group, they drive the size of the total electorate higher, meaning you need more alive members to reach quorum. The result is elections that fail to complete, for lack of quorum, when they "should" complete from a human perspective.

# tl;dr - the fix

There is no automatic way to resolve this use case. We propose implementing a command that allows an administrator to inform a service group that it should mark all `confirmed` members of the service group as no longer part of the electorate. 

# The details

This quorum is calculated with two inputs - _the electorate_ (the total number of members who have the service group running) and the _alive members_ (the number of members who have the service group running and are also marked as 'alive'). 

To understand the problem, lets start with the minimum sized valid electorate of 3:

| Member | Status | Leader |
|---|---|---|
| 1 | Alive | N |
| 2 | Alive | N |
| 3 | Alive | Y |

Now, lets say member 3 goes down:

| Member | Status | Leader |
|---|---|---|
| 1 | Alive | N |
| 2 | Alive | N |
| 3 | Confirmed | NA |

Under the hood, we have an _electorate_ of 3, and an _alive-members_ of 2. The formula for a quorum is `alive_members >= ((electorate/2)+1)` - so we wind up with an election, and the state is now:

| Member | Status | Leader |
|---|---|---|
| 1 | Alive | N |
| 2 | Alive | Y |
| 3 | Confirmed | NA |

Now, time goes by - and our administration team helpfully launches another member to take the place of dear-departed member 3.

| Member | Status | Leader |
|---|---|---|
| 1 | Alive | N |
| 2 | Alive | Y |
| 3 | Confirmed | NA |
| 4 | Alive | N |

Suddenly, the difficulty of being a leader amongst the ruffians in the service group gets to be too much for member 2:

| Member | Status | Leader |
|---|---|---|
| 1 | Alive | N |
| 2 | Confirmed | NA |
| 3 | Confirmed | NA |
| 4 | Alive | N |

So we now have `2 >= ((4/2)+1)`, or `2 >= 3`, and we fail to achieve quorum - there is no way to know if the condition we are in is a 2x2 net split . So we add a few more members, because we don't want to be in this situation again:

| Member | Status | Leader |
|---|---|---|
| 1 | Alive | Y |
| 2 | Confirmed | NA |
| 3 | Confirmed | NA |
| 4 | Alive | N |
| 5 | Alive | N |
| 6 | Alive | N |
| 7 | Alive | N |

Everything is fine - 5 >= 4, so we have quorum.  Suddenly, a we loose 4 members:

| Member | Status | Leader |
|---|---|---|
| 1 | Confirmed | NA |
| 2 | Confirmed | NA |
| 3 | Confirmed | NA |
| 4 | Confirmed | NA |
| 5 | Confirmed | NA |
| 6 | Confirmed | NA |
| 7 | Alive | N |

Thinking we know what to do, we add 4 more:

| Member | Status | Leader |
|---|---|---|
| 1 | Confirmed | NA |
| 2 | Confirmed | NA |
| 3 | Confirmed | NA |
| 4 | Confirmed | NA |
| 5 | Confirmed | NA |
| 6 | Confirmed | NA |
| 7 | Alive | N |
| 8 | Alive | N |
| 9 | Alive | N |
| 10 | Alive | N |
| 11 | Alive | N |

And arithmetic rears its ugly head - `5 >= 6`, and we fail again. And so it goes. Obviously, if we kept trying to just have a cluster size of 3, this happens much sooner.

# Mitigations we discussed but dismissed

## Persistent Supervisors

The work to have supervisors be persistent helps here, in that it's much harder to get into this situation if the members are actually coming back. They'll pick up where they left off, mark themselves alive again, and the electorate doesn't inflate arbitrarily. It doesn't solve the problem, though - as members die permanently, you'll eventually have to raise the size of the service group to keep up.

## Garbage Collection

We can garbage collect confirmed members. This is probably a good idea, but the question here is - after how long? Any aggressive setting runs the risk of creating a scenario where short-lived net splits protect you from a split brain, but long-lived ones (longer than whatever your threshold is) do not. This is unacceptable, since a split-brain scenario is the worst case. I think we probably do implement garbage collection, but on a threshold somewhere in the 72-96 hour range.

# The proposed fix

We can add a mechanism for informing a service group that they should consider all currently `confirmed` members as no longer valid parts of the electorate. Upon receipt of the rumor, a member would mark all of its confirmed members as having service entires that are no longer part of the electorate.  

`hab sup mark confirmed redis.default`

# In the meantime

We recommend using the persistent supervisors when in concert with the leader topology, and working hard to bring them back from the dead.

# Thanks

This issue was first recognized by @moretea, who helpfully created a video and a gist on how to reproduce:

https://gist.github.com/moretea/f69f71b342a1d10d7ee05e52dafffd6a
https://www.youtube.com/watch?v=cVIH87B-ZO0

Thanks!

**Comment by adamhjk** [comment]
I have verified that this would happen with both `etcd` and `zookeeper` as well - if you add members to the cluster without removing them, you'll quickly loose quorum. In both cases, the answer is to manually inform the cluster that those nodes are gone and not coming back.

**Comment by adamhjk** [comment]
Note that in both those cases, this is different from wether this would be a problem within an election *enabled* by etcd or zookeeper.  In both those cases, the election operation is usually just a race to win an atomic compare-and-swap.

**Comment by mwrock** [comment]
Thanks for the detailed explanation above! It made this scenario easy to reproduce in my head.

I think adding a cli option to manually "prune" the electorate and a super conservative default maximum time confirmed before auto removal is reasonable.

**Comment by bodymindarts** [comment]
I think when contemplating desired behaviour we need to think about 3 scenarios:

1) Node purposefully leaves cluster due to downscaling or maintenance. https://github.com/habitat-sh/habitat/issues/1812 already is an open ticket around this. Consul has a nice way of handling this via a (`leave_on_terminate`](https://www.consul.io/docs/agent/options.html#leave_on_terminate) setting that will cause a server to gracefully leave the cluster when the it receives a TERM signal.
2) There is a network partition or other kind of connectivity issue. Here a node may come back but we do not know when.
3) A node has crashed and will never return. Perhaps the underlying scheduler brings up another one as a replacement.

Differentiating between 2 and 3 is where we need to be carefull. Consul has the [reconnect_timeout](https://www.consul.io/docs/agent/options.html#reconnect_timeout) that defaults to 72hours (minimum 8hours) and the [force-leave](https://www.consul.io/docs/commands/force-leave.html) command to manually mark failed nodes as gone.

Looks like the proposed solution is a reasonable approach.

**Comment by adamhjk** [comment]
Agreed - we also need `leave_on_terminate`, precisely for the #1812 use case. 

On implementation, I say we do the equivalent of `force-leave` first, and then we do the timeout.

**Comment by adamhjk** [comment]
For any future implementors:

1. Members need to have a 'departed' field added.
1. On graceful shutdown, we set our member rumor to departed and share the rumor somewhat aggressively. This will delay our shutdown somewhat.
1. Any member, departed or not, should be garbage collected after 72 hours.
1. Any member can be forced to depart with `hab-butterfly member depart MEMBERID`
1. When a quorum fails to be reached, we should print a message telling you what members you could force to depart to restore quorum.

**Comment by adamhjk** [comment]
Consul just implemented a clever solution here. If members have failed, and a new member joins, we force a dead member out of quorum. The side effect is auto scale and remediation works as expected, as do net splits. Clever!

**Comment by moretea** [comment]
> Consul just implemented a clever solution here. If members have failed, and a new member joins, we force a dead member out of quorum. The side effect is auto scale and remediation works as expected, as do net splits. Clever!

What's the time out there? I assume it's less than 72 hours. If you would do a rolling upgrade of e.g. your CoreOS cluster, you could get into trouble quite easily if the members in the quorum are GC'ed after 72 hours.

**Comment by bodymindarts** [comment]
@moretea a rolling upgrade is not the same as a node failing. As long as a node retains its `MemberId` which it does during a service upgrade it is still considered the same node from the POV of habitat.

The 72 hours is for a node that goes down due to an infrastructure failure and never returns.

**Comment by moretea** [comment]
@bodymindarts Ah yes, I was talking about running supervisors on top of an orchestrator such as Kubernetes. Pods are (and should be!) be ephemeral, and therefore it is impossible to retain a `MemberID` over restarts. Having a timeout of 72 hours might make sense for cases where manual labour is involved to get a physical machine up and running again, but in cloud native systems, one should simply discard the old supervisor and create a new one.

> If a human operator needs to touch your system during normal operations, you have a bug. The definition of normal changes as your systems grow.
https://landing.google.com/sre/book/chapters/eliminating-toil.html

**Comment by bodymindarts** [comment]
@moretea in the case where you let the scheduler handle the update you need a way to deliberately and gracefully leave the cluster and the scheduler has to signal the supervisor that that is what is happening.
Then the supervisor will gracefully leave the cluster and you won't have the issue.
Does K8s have that kind of mechanism? Some kind of pre-delete hook on a pod?

**Comment by bodymindarts** [comment]
Also the thing that Adam mentions is instantaneous. If there is a dead node in the gossip ring and another one comes up it should replace the old one effective immediately. (I think this is how it is in the latest consul version.

**Comment by moretea** [comment]
@bodymindarts yes, kubernetes will send a `TERM` signal, which could be used to leave the quorum gracefully.

**Comment by bodymindarts** [comment]
@moretea exactly that is what we should be doing on `TERM`.

**Comment by moretea** [comment]
OK cool, if I have time I'll investigate this new behavior.

**Comment by bodymindarts** [comment]
It isn't implemented yet. This ticket is exactly that issue (graceful shutdown).

**Comment by adamhjk** [comment]
For anyone who comes later - this was fixed by adding "departure" to the protocol, and auto-departing members when new members join a service group.
